samples/firefox-accounts/src/main/java/org/mozilla/samples/fxa/MainActivity.kt

```diff
@@ -37,44 +38,50 @@ open class MainActivity : AppCompatActivity(), LoginFragment.OnLoginCompleteList
         super.onCreate(savedInstanceState)
         setContentView(R.layout.activity_main)
 
-        getSharedPreferences(FXA_STATE_PREFS_KEY, Context.MODE_PRIVATE).getString(FXA_STATE_KEY, "").let {
-            FirefoxAccount.fromJSONString(it).then({
-                this.account = it
-                account?.getProfile()
-            }, {
-                getIntent().getExtras()?.getString("pairingUrl")?.let {
-                    Config.custom(CONFIG_URL_PAIRING).then { value: Config ->
-                        val acct = FirefoxAccount(value, CLIENT_ID, REDIRECT_URL)
-                        account = acct
-
-                        account?.beginPairingFlow(it, scopes)?.whenComplete {
-                            openWebView(it)
-                        }
+        whenAccount = async {
+            try {
+                val savedJSON = getSharedPreferences(FXA_STATE_PREFS_KEY, Context.MODE_PRIVATE)
+                        .getString(FXA_STATE_KEY, "")
+                val acct = FirefoxAccount.fromJSONString(savedJSON).await()
 
-                        acct.getProfile()
+                val profile = whenAccount.await().getProfile(true).await()
+                val txtView: TextView = findViewById(R.id.txtView)
+                launch(UI) {
+                    txtView.text = getString(R.string.signed_in,
+                            "${profile.displayName ?: ""} ${profile.email}")
+                }
+                acct
+            } catch (e: FxaException) {
```

Comment:
It's not ideal that we handle the main auth flow in an Exception block. I think that's something we should fix, but we can file a follow-up issue if that's a big change right now.

Reply:
I think this should be cleaned up in a follow up.

Reply:
Follow-up filed: https://github.com/mozilla-mobile/android-components/issues/1230